File: tasks/native-runtime/file_cache_parity_probe.py
"""File-cache and projects-branch parity probe, Python side (slice E2).

Covers `load_cached_file` and its helpers from `infra/rag/files.py`, plus the
`list_project_files_tool` / `read_file_chunk_tool` / `project_document_for_tool`
bodies from `tool_runtime/tools.py`, on top of the projects store already verified in
`projects_parity_probe.py` (imported here, not re-extracted).

Usage::

    python tasks/native-runtime/file_cache_parity_probe.py > python.json
    cd rust && cargo run -p deepseek-policy --example file_cache_parity_probe > ../rust.json
    diff <(tr -d '\\r' < python.json) <(tr -d '\\r' < rust.json)
"""

from __future__ import annotations

import ast
import json
import shutil
import sys
import tempfile
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
import projects_parity_probe as projects_probe  # noqa: E402

REPO = Path(__file__).resolve().parents[2]
FILES = REPO / "deepseek_infra" / "infra" / "rag" / "files.py"
TOOLS = REPO / "deepseek_infra" / "infra" / "tool_runtime" / "tools.py"
ERRORS = REPO / "deepseek_infra" / "core" / "errors.py"

FILE_FUNCTIONS = (
    "load_cached_file",
    "_load_cached_file_cached",
    "_load_cached_file_impl",
    "_load_cached_file_impl_from_path",
    "project_file_cache_dir",
)
TOOL_FUNCTIONS = (
    "list_project_files_tool",
    "project_document_for_tool",
    "read_file_chunk_tool",
)

GOOD_ID = "a" * 32

INDEX_BODY = {
    "id": "file-under-test",
    "name": "Report.pdf",
    "kind": "pdf",
    "projectId": "abcd",
    "chunks": [
        {"lineStart": 1, "lineEnd": 10, "text": "first chunk"},
        {"lineStart": 11, "lineEnd": 20, "text": "x" * 7000},
        {"lineStart": 21, "lineEnd": 30, "text": 42},
        "not-a-chunk",
    ],
}

LOAD_CASES = [
    ("bad-short-id", "abc", None),
    ("bad-uppercase-id", "A" * 32, None),
    ("bad-nonhex-id", "g" * 32, None),
    ("empty-id", "", None),
]

CHUNK_CASES = [
    ("first-default", {}, None),
    ("index-one", {"chunkIndex": 1}, None),
    ("index-two", {"chunkIndex": 2}, None),
    ("index-last", {"chunkIndex": 4}, None),
    ("out-of-range", {"chunkIndex": 5}, None),
    ("large-index", {"chunkIndex": 99}, None),
    ("zero-index", {"chunkIndex": 0}, None),
    ("negative-index", {"chunkIndex": -1}, None),
    ("non-dict-chunk", {"chunkIndex": 4}, None),
    ("missing-chunks", {}, "no-chunks"),
    ("chunks-not-list", {}, "chunks-scalar"),
    ("project-scoped", {"chunkIndex": 1, "projectId": "abcd"}, "project"),
    ("bad-project-id", {"chunkIndex": 1, "projectId": "ab"}, None),
]


def _extract(source: str, name: str) -> str:
    for node in ast.walk(ast.parse(source)):
        if isinstance(node, ast.FunctionDef) and node.name == name:
            segment = ast.get_source_segment(source, node)
            if node.decorator_list:
                decorators = "\n".join(
                    "@" + ast.unparse(decorator) for decorator in node.decorator_list
                )
                return decorators + "\n" + segment
            return segment
    raise SystemExit(f"could not extract {name}")


def build_namespace(root: Path) -> dict:
    # The projects store, from its own probe.
    namespace = projects_probe.build_namespace(root)

    import functools
    import re
    from pathlib import Path as _Path
    from typing import Any

    file_cache_dir = root / ".file-cache"
    namespace.update(
        {
            "FILE_CACHE_DIR": file_cache_dir,
            "PROJECTS_DIR": root / ".projects",
            "FILE_READER_MAX_CHUNKS": 12,
            "FILE_SOURCE_SUFFIX": ".source",
            "functools": functools,
            "lru_cache": functools.lru_cache,
            "_Path": _Path,
            "Any": Any,
            "re": re,
        }
    )

    files_source = FILES.read_text(encoding="utf-8")
    for name in FILE_FUNCTIONS:
        exec(compile(_extract(files_source, name), str(FILES), "exec"), namespace)  # noqa: S102

    tools_source = TOOLS.read_text(encoding="utf-8")
    for name in TOOL_FUNCTIONS:
        exec(compile(_extract(tools_source, name), str(TOOLS), "exec"), namespace)  # noqa: S102
    return namespace


def outcome(call) -> dict:
    try:
        return {"ok": True, "result": call()}
    except Exception as exc:  # noqa: BLE001 - the probe reports the shape
        return {
            "ok": False,
            "error": str(exc),
            "code": getattr(getattr(exc, "code", None), "value", None),
            "status": getattr(exc, "status", None),
        }


def main() -> int:
    for path in (FILES, TOOLS, ERRORS):
        if not path.exists():
            print(f"missing {path}", file=sys.stderr)
            return 2

    root = Path(tempfile.mkdtemp(prefix="file-cache-parity-"))
    out: dict = {}
    try:
        ns = build_namespace(root)
        cache_dir = ns["FILE_CACHE_DIR"]
        project_files = ns["PROJECTS_DIR"] / "abcd" / "files"

        def write_index(directory: Path, body) -> None:
            directory.mkdir(parents=True, exist_ok=True)
            text = body if isinstance(body, str) else json.dumps(body, ensure_ascii=False, indent=2)
            (directory / f"{GOOD_ID}.json").write_text(text, encoding="utf-8")

        # --- load_cached_file error shapes ------------------------------------
        for label, file_id, _ in LOAD_CASES:
            out[f"load::{label}"] = outcome(lambda file_id=file_id: ns["load_cached_file"](file_id))

        out["load::missing"] = outcome(lambda: ns["load_cached_file"](GOOD_ID))

        write_index(cache_dir, "{not json")
        out["load::malformed"] = outcome(lambda: ns["load_cached_file"](GOOD_ID))
        write_index(cache_dir, "42")
        out["load::scalar"] = outcome(lambda: ns["load_cached_file"](GOOD_ID))

        # --- the project-scoped path ------------------------------------------
        write_index(project_files, INDEX_BODY)
        out["load::global-cannot-see-project"] = outcome(lambda: ns["load_cached_file"](GOOD_ID))
        out["load::scoped"] = outcome(lambda: ns["load_cached_file"](GOOD_ID, project_id="abcd"))
        out["load::bad-project-id"] = outcome(
            lambda: ns["load_cached_file"](GOOD_ID, project_id="ab")
        )
        out["load::blank-project-id-is-global"] = outcome(
            lambda: ns["load_cached_file"](GOOD_ID, project_id="   ")
        )
        out["cache-dir::project"] = outcome(
            lambda: str(ns["project_file_cache_dir"]("abcd").relative_to(root))
        )

        # --- read_file_chunk_tool ---------------------------------------------
        def serve(body=INDEX_BODY, scoped=False) -> None:
            target = project_files if scoped else cache_dir
            write_index(target, body)

        for label, arguments, variant in CHUNK_CASES:
            body = INDEX_BODY
            scoped = False
            if variant == "no-chunks":
                body = {"id": "f", "name": "n", "kind": "text"}
            elif variant == "chunks-scalar":
                body = {"id": "f", "name": "n", "kind": "text", "chunks": "no"}
            elif variant == "project":
                scoped = True
            serve(body, scoped=scoped)
            if scoped:
                # The global copy must not exist, so only the project path can serve.
                (cache_dir / f"{GOOD_ID}.json").unlink(missing_ok=True)
            else:
                write_index(cache_dir, body)
                (project_files / f"{GOOD_ID}.json").unlink(missing_ok=True)
            out[f"chunk::{label}"] = outcome(
                lambda arguments=arguments: ns["read_file_chunk_tool"](
                    str(arguments.get("fileId") or GOOD_ID),
                    chunk_index=int(arguments.get("chunkIndex") or 0),
                    project_id=str(arguments.get("projectId") or ""),
                )
            )

        # --- list_project_files_tool ------------------------------------------
        def write_project(project_id: str, body: dict) -> None:
            target = ns["PROJECTS_DIR"] / project_id
            target.mkdir(parents=True, exist_ok=True)
            (target / "project.json").write_text(
                json.dumps(body, ensure_ascii=False, indent=2), encoding="utf-8"
            )

        documents = [
            {"fileId": "b" * 32, "projectId": "abcd", "name": "A", "kind": "pdf",
             "pageCount": 2, "charCount": 30, "chunkCount": 4, "preview": "p" * 700},
            {"fileId": "c" * 32, "projectId": "abcd", "name": "B"},
            {"fileId": "bad", "projectId": "abcd"},
            "not-a-dict",
        ] + [
            {"fileId": f"{index:032x}", "projectId": "abcd", "name": f"D{index}"}
            for index in range(200)
        ]
        write_project("abcd", {"name": "Alpha", "documents": documents, "updatedAt": 5})
        write_project("efgh", {"name": "Beta", "documents": [], "updatedAt": 9})

        out["list::named"] = outcome(lambda: ns["list_project_files_tool"]("abcd"))
        out["list::missing"] = outcome(lambda: ns["list_project_files_tool"]("zzzz"))
        out["list::invalid-id"] = outcome(lambda: ns["list_project_files_tool"]("ab"))
        listed = outcome(lambda: ns["list_project_files_tool"](""))
        # The full payload is large; compare the shape and the counts.
        if listed.get("ok"):
            payload = listed["result"]
            out["list::all"] = {
                "ok": True,
                "ids": [project["id"] for project in payload["projects"]],
                "count": payload["count"],
                "per-project": [len(project["files"]) for project in payload["projects"]],
                "first-file": next(
                    (project["files"][0] for project in payload["projects"] if project["files"]),
                    None,
                ),
            }
        else:
            out["list::all"] = listed
    finally:
        shutil.rmtree(root, ignore_errors=True)

    json.dump(out, sys.stdout, ensure_ascii=False, indent=2, sort_keys=True)
    sys.stdout.write("\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: rust/crates/deepseek-policy/examples/file_cache_parity_probe.rs
//! File-cache and projects-branch parity probe, Rust side (slice E2).
//!
//! Replays the same corpus as `tasks/native-runtime/file_cache_parity_probe.py`
//! through `deepseek_policy::file_cache` and `deepseek_policy::projects`.
//!
//! Usage::
//!
//!     python tasks/native-runtime/file_cache_parity_probe.py > python.json
//!     cd rust && cargo run -p deepseek-policy --example file_cache_parity_probe > ../rust.json
//!     diff <(tr -d '\r' < python.json) <(tr -d '\r' < rust.json)

use std::fs;
use std::path::{Path, PathBuf};
use std::sync::atomic::{AtomicU64, Ordering};

use deepseek_policy::app_error::AppError;
use deepseek_policy::entropy::Entropy;
use deepseek_policy::file_cache::{
    FileCache, file_cache_dir, load_cached_file, project_file_cache_dir,
};
use deepseek_policy::projects::{list_project_files, projects_dir, read_file_chunk};
use serde_json::{Map, Value, json};

const GOOD_ID: &str = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa";

struct Scratch {
    root: PathBuf,
}

impl Scratch {
    fn new(label: &str) -> Self {
        let root =
            std::env::temp_dir().join(format!("file-cache-parity-{label}-{}", std::process::id()));
        let _ = fs::remove_dir_all(&root);
        fs::create_dir_all(&root).expect("create scratch root");
        Self { root }
    }
    fn path(&self) -> &Path {
        &self.root
    }
}

impl Drop for Scratch {
    fn drop(&mut self) {
        let _ = fs::remove_dir_all(&self.root);
    }
}

struct FixedEntropy {
    ids: AtomicU64,
}

impl Entropy for FixedEntropy {
    fn new_id(&self) -> Result<String, AppError> {
        let value = self.ids.fetch_add(1, Ordering::Relaxed) + 1;
        Ok(format!("{value:016x}"))
    }
    fn now_millis(&self) -> i64 {
        1_760_000_000_000
    }
}

/// The index the Python probe serves. `chunks[3]` is deliberately not an object.
fn index_body() -> Value {
    json!({
        "id": "file-under-test",
        "name": "Report.pdf",
        "kind": "pdf",
        "projectId": "abcd",
        "chunks": [
            {"lineStart": 1, "lineEnd": 10, "text": "first chunk"},
            {"lineStart": 11, "lineEnd": 20, "text": "x".repeat(7000)},
            {"lineStart": 21, "lineEnd": 30, "text": 42},
            "not-a-chunk"
        ]
    })
}

fn chunk_cases() -> Vec<(&'static str, Value, &'static str)> {
    vec![
        ("first-default", json!({}), "normal"),
        ("index-one", json!({"chunkIndex": 1}), "normal"),
        ("index-two", json!({"chunkIndex": 2}), "normal"),
        ("index-last", json!({"chunkIndex": 4}), "normal"),
        ("out-of-range", json!({"chunkIndex": 5}), "normal"),
        ("large-index", json!({"chunkIndex": 99}), "normal"),
        ("zero-index", json!({"chunkIndex": 0}), "normal"),
        ("negative-index", json!({"chunkIndex": -1}), "normal"),
        ("non-dict-chunk", json!({"chunkIndex": 4}), "normal"),
        ("missing-chunks", json!({}), "no-chunks"),
        ("chunks-not-list", json!({}), "chunks-scalar"),
        (
            "project-scoped",
            json!({"chunkIndex": 1, "projectId": "abcd"}),
            "project",
        ),
        (
            "bad-project-id",
            json!({"chunkIndex": 1, "projectId": "ab"}),
            "normal",
        ),
    ]
}

fn outcome<T: serde::Serialize>(result: Result<T, AppError>) -> Value {
    match result {
        Ok(value) => json!({"ok": true, "result": value}),
        Err(error) => json!({
            "ok": false,
            "error": error.message,
            "code": error.code,
            "status": error.status,
        }),
    }
}

fn write_index(directory: &Path, body: &Value) {
    fs::create_dir_all(directory).expect("create cache dir");
    let text = match body {
        Value::String(raw) => raw.clone(),
        other => serde_json::to_string_pretty(other).expect("serialize index"),
    };
    fs::write(directory.join(format!("{GOOD_ID}.json")), text).expect("write index");
}

fn write_project(root: &Path, project_id: &str, body: &Value) {
    let target = projects_dir(root).join(project_id);
    fs::create_dir_all(&target).expect("create project dir");
    fs::write(
        target.join("project.json"),
        serde_json::to_string_pretty(body).expect("serialize project"),
    )
    .expect("write project.json");
}

fn main() {
    let scratch = Scratch::new("rust");
    let root = scratch.path();
    let entropy = FixedEntropy {
        ids: AtomicU64::new(0),
    };
    let cache = FileCache::new();
    let mut out = Map::new();

    let global_dir = file_cache_dir(root);
    let project_files = project_file_cache_dir(root, "abcd").expect("project files dir");

    // --- load_cached_file error shapes ----------------------------------------
    for (label, file_id) in [
        ("bad-short-id", "abc"),
        ("bad-uppercase-id", "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"),
        ("bad-nonhex-id", "gggggggggggggggggggggggggggggggg"),
        ("empty-id", ""),
    ] {
        out.insert(
            format!("load::{label}"),
            outcome(load_cached_file(root, file_id, None, &cache)),
        );
    }
    out.insert(
        "load::missing".to_string(),
        outcome(load_cached_file(root, GOOD_ID, None, &cache)),
    );

    write_index(&global_dir, &json!("{not json"));
    out.insert(
        "load::malformed".to_string(),
        outcome(load_cached_file(root, GOOD_ID, None, &cache)),
    );
    write_index(&global_dir, &json!("42"));
    out.insert(
        "load::scalar".to_string(),
        outcome(load_cached_file(root, GOOD_ID, None, &cache)),
    );

    // --- the project-scoped path ----------------------------------------------
    write_index(&project_files, &index_body());
    out.insert(
        "load::global-cannot-see-project".to_string(),
        outcome(load_cached_file(root, GOOD_ID, None, &cache)),
    );
    out.insert(
        "load::scoped".to_string(),
        outcome(load_cached_file(root, GOOD_ID, Some("abcd"), &cache)),
    );
    out.insert(
        "load::bad-project-id".to_string(),
        outcome(load_cached_file(root, GOOD_ID, Some("ab"), &cache)),
    );
    // A whitespace-only id is truthy, so it reaches the shape check and fails.
    out.insert(
        "load::blank-project-id-is-global".to_string(),
        outcome(load_cached_file(root, GOOD_ID, Some("   "), &cache)),
    );
    out.insert(
        "cache-dir::project".to_string(),
        outcome(project_file_cache_dir(root, "abcd").map(|path| {
            path.strip_prefix(root)
                // Python's `str(Path)` uses the native separator, so it is not
                // normalised away here.
                .map(|relative| relative.to_string_lossy().to_string())
                .unwrap_or_default()
        })),
    );

    // --- read_file_chunk_tool -------------------------------------------------
    for (label, arguments, variant) in chunk_cases() {
        let scoped = variant == "project";
        let body = match variant {
            "no-chunks" => json!({"id": "f", "name": "n", "kind": "text"}),
            "chunks-scalar" => json!({"id": "f", "name": "n", "kind": "text", "chunks": "no"}),
            _ => index_body(),
        };
        if scoped {
            write_index(&project_files, &body);
            // The global copy must not exist, so only the project path can serve.
            let _ = fs::remove_file(global_dir.join(format!("{GOOD_ID}.json")));
        } else {
            write_index(&global_dir, &body);
            let _ = fs::remove_file(project_files.join(format!("{GOOD_ID}.json")));
        }
        // The probe's dispatch falls back to GOOD_ID when `fileId` is absent.
        let mut served = arguments.as_object().cloned().unwrap_or_default();
        if !served.contains_key("fileId") {
            served.insert("fileId".to_string(), json!(GOOD_ID));
        }
        out.insert(
            format!("chunk::{label}"),
            outcome(read_file_chunk(&served, root, &cache)),
        );
    }

    // --- list_project_files_tool ----------------------------------------------
    let mut documents: Vec<Value> = vec![
        json!({"fileId": "b".repeat(32), "projectId": "abcd", "name": "A", "kind": "pdf",
               "pageCount": 2, "charCount": 30, "chunkCount": 4, "preview": "p".repeat(700)}),
        json!({"fileId": "c".repeat(32), "projectId": "abcd", "name": "B"}),
        json!({"fileId": "bad", "projectId": "abcd"}),
        json!("not-a-dict"),
    ];
    for index in 0..200 {
        documents.push(json!({
            "fileId": format!("{index:032x}"),
            "projectId": "abcd",
            "name": format!("D{index}")
        }));
    }
    write_project(
        root,
        "abcd",
        &json!({"name": "Alpha", "documents": documents, "updatedAt": 5}),
    );
    write_project(
        root,
        "efgh",
        &json!({"name": "Beta", "documents": [], "updatedAt": 9}),
    );

    let named = |project_id: &str| {
        let mut arguments = Map::new();
        arguments.insert("projectId".to_string(), json!(project_id));
        outcome(list_project_files(&arguments, root, &entropy))
    };
    out.insert("list::named".to_string(), named("abcd"));
    out.insert("list::missing".to_string(), named("zzzz"));
    out.insert("list::invalid-id".to_string(), named("ab"));

    let listed = outcome(list_project_files(&Map::new(), root, &entropy));
    out.insert(
        "list::all".to_string(),
        match &listed {
            Value::Object(fields) if fields.get("ok") == Some(&Value::Bool(true)) => {
                let payload = &fields["result"];
                let projects = payload["projects"].as_array().cloned().unwrap_or_default();
                let ids: Vec<Value> = projects
                    .iter()
                    .map(|project| project["id"].clone())
                    .collect();
                let per_project: Vec<Value> = projects
                    .iter()
                    .map(|project| json!(project["files"].as_array().map(Vec::len).unwrap_or(0)))
                    .collect();
                let first_file = projects
                    .iter()
                    .find_map(|project| {
                        project["files"]
                            .as_array()
                            .and_then(|files| files.first().cloned())
                    })
                    .unwrap_or(Value::Null);
                json!({
                    "ok": true,
                    "ids": ids,
                    "count": payload["count"],
                    "per-project": per_project,
                    "first-file": first_file,
                })
            }
            _ => listed,
        },
    );

    let mut encoded =
        serde_json::to_string_pretty(&Value::Object(out)).expect("serialize probe output");
    encoded.push('\n');
    print!("{encoded}");
}
